distributed/utils.py

```diff
@@ -470,62 +518,53 @@ def start(self):
+        if self._loop is not None:
+            try:
+                count, real_runner = self._all_loops[self._loop]
+            except KeyError:
+                assert self._loop.asyncio_loop.is_running()  # type: ignore[attr-defined]
+                self._started = True
+                return
+
             self._all_loops[self._loop] = count + 1, real_runner
             self._started = True
             return
 
         assert self._loop_thread is None
-        assert count == 0
 
-        loop_evt = threading.Event()
-        done_evt = threading.Event()
-        in_thread = [None]
-        start_exc = [None]
+        start_evt = threading.Event()
+        start_exc = None
+        loop = None
 
-        def loop_cb():
-            in_thread[0] = threading.current_thread()
-            loop_evt.set()
+        async def amain() -> None:
+            nonlocal loop
+            loop = IOLoop.current()
+            start_evt.set()
+            await self._stop_event.wait()
 
-        def run_loop(loop=self._loop):
-            loop.add_callback(loop_cb)
-            # run loop forever if it's not running already
+        def run_loop() -> None:
+            nonlocal start_exc
             try:
-                if not loop.asyncio_loop.is_running():
-                    loop.start()
-            except Exception as e:
-                start_exc[0] = e
-            finally:
-                done_evt.set()
-
-        thread = threading.Thread(target=run_loop, name="IO loop")
-        thread.daemon = True
-        thread.start()
-
-        loop_evt.wait(timeout=10)
+                asyncio.run(amain())
+            except BaseException as e:
+                if start_evt.is_set():
+                    raise
+                start_exc = e
+                start_evt.set()
+
+        self._loop_thread = _CollectErrorThread(
+            target=run_loop, daemon=True, name="IO loop"
+        )
+        start_evt.wait(timeout=10)
+        if start_exc is not None:
+            raise start_exc
+        assert loop is not None
+        self._loop = loop
         self._started = True
-
-        actual_thread = in_thread[0]
-        if actual_thread is not thread:
-            # Loop already running in other thread (user-launched)
-            done_evt.wait(5)
-            if start_exc[0] is not None and not isinstance(start_exc[0], RuntimeError):
-                if not isinstance(
-                    start_exc[0], Exception
-                ):  # track down infrequent error
-                    raise TypeError(
-                        f"not an exception: {start_exc[0]!r}",
-                    )
-                raise start_exc[0]
-            self._all_loops[self._loop] = count + 1, None
-        else:
-            assert start_exc[0] is None, start_exc
-            self._loop_thread = thread
-            self._all_loops[self._loop] = count + 1, self
+        self._all_loops[loop] = (1, self)
```

Comment:
I think this is an absurd case to worry about, but is it technically possible to somehow have multiple `LoopRunner`s starting in parallel in multiple threads, but targeting the same `loop`, and then overwriting each other here?

Reply:
This is run protected by a lock